{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e3b8ce8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from itertools import chain\n",
    "import csv\n",
    "from collections import defaultdict\n",
    "import plotly.express as px\n",
    "import glob\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06dc8893",
   "metadata": {},
   "source": [
    "### Read in databases for DR, repeats, NE genes and lineages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "adafe06a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drug resistance positions\n",
    "DB_DR = []\n",
    "df = pd.read_csv('bin/db/drdb.txt',sep='\\t')\n",
    "df = df[df['Reference_Position'] != '-']\n",
    "for snp in df['Reference_Position']:\n",
    "    snp = snp.split('/')\n",
    "    for i in snp:\n",
    "        DB_DR.append(float(i))\n",
    "        \n",
    "# repeat positions and ne genes\n",
    "df1 = pd.read_csv('bin/db/repeats.txt',sep='\\t')\n",
    "df2 = pd.read_csv('bin/db/ne_genes.txt',sep='\\t')\n",
    "df = pd.concat([df1,df2])\n",
    "DB_REP_NE = df.apply(lambda row:list(range(row['Start'],row['End']+1)),axis=1).explode().astype(int).tolist()\n",
    "\n",
    "# lineages\n",
    "df_lin = pd.read_csv('bin/db/lineages.bed',sep='\\t',index_col=1)\n",
    "DB_LIN = df_lin.drop(columns=['Chromosome','End'])\n",
    "colors = px.colors.qualitative.Dark24\n",
    "colors.append(px.colors.qualitative.Light24[1])\n",
    "colors = dict(zip(DB_LIN['Lineage_name'].unique(),colors))\n",
    "DB_LIN['Lineage_color'] = DB_LIN['Lineage_name'].map(colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a317acce",
   "metadata": {},
   "source": [
    "### Set global variables and read in files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "de94d191",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SET GLOBAL VARIABLES HERE\n",
    "FREQ = 90\n",
    "COV = 10\n",
    "QUAL = 20\n",
    "FRB = 0\n",
    "\n",
    "OUTPUT = 'output/'\n",
    "INPUT = 'input/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c8d013e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename files\n",
    "directory = Path(INPUT+\"snp/\")\n",
    "files = [f for f in directory.iterdir() if f.is_file()]\n",
    "for old_path in files:\n",
    "    dir_name = os.path.dirname(old_path)\n",
    "    base_name = os.path.basename(old_path)\n",
    "    if os.path.splitext(base_name)[1] == '.csv':\n",
    "        new_base = base_name.split('_')[0] + '.snp'\n",
    "        new_path = os.path.join(dir_name, new_base)\n",
    "        os.rename(old_path, new_path)\n",
    "    \n",
    "directory = Path(INPUT+\"cov/\")\n",
    "files = [f for f in directory.iterdir() if f.is_file()]\n",
    "for old_path in files:\n",
    "    dir_name = os.path.dirname(old_path)\n",
    "    base_name = os.path.basename(old_path)\n",
    "    if os.path.splitext(base_name)[1] == '.csv':\n",
    "        new_base = base_name.split('_')[0] + '.cov'\n",
    "        new_path = os.path.join(dir_name, new_base)\n",
    "        os.rename(old_path, new_path)\n",
    "    \n",
    "directory = Path(INPUT+\"stat/\")\n",
    "files = [f for f in directory.iterdir() if f.is_file()]\n",
    "for old_path in files:\n",
    "    dir_name = os.path.dirname(old_path)\n",
    "    base_name = os.path.basename(old_path)\n",
    "    if os.path.splitext(base_name)[1] == '.csv':\n",
    "        new_base = base_name.split('_')[0] + '.stat'\n",
    "        new_path = os.path.join(dir_name, new_base)\n",
    "        os.rename(old_path, new_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f5080a74",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_files = os.listdir(INPUT+'snp/')\n",
    "cov_files = os.listdir(INPUT+'cov/')\n",
    "stat_files = os.listdir(INPUT+'stat/')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f56f2c1",
   "metadata": {},
   "source": [
    "### Generate a reference dictionary for all SNPs that passed QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "573c19df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TO DO:\n",
    "\n",
    "# --> average coverage check using cov files before hand (in snp2tree this is pass_cov_check) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "73cf3d78",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TC00760619_S1.snp\n",
      "TC00856554_S51.snp\n",
      "TC00764189.snp\n",
      "TC00831960_S32.snp\n",
      "TC00855201_S148.snp\n",
      "TC00862935_S189.snp\n",
      "TC00836558_1.snp\n",
      "TC00777418_S47.snp\n",
      "TC00832902_S78.snp\n",
      "TC00780707_S93.snp\n",
      "TC00838386_S26.snp\n"
     ]
    }
   ],
   "source": [
    "# this will happen after QC of filenames etc and just the final files in the input\n",
    "\n",
    "ref_dict = dict()\n",
    "snp_dfs = dict()\n",
    "snp_dfs_filtered = dict()\n",
    "\n",
    "for file in snp_files:\n",
    "    print(file)\n",
    "    sample = file.replace('.snp','')\n",
    "    df = pd.read_csv('input/snp/'+file)\n",
    "    snp_dfs[sample] = df\n",
    "    \n",
    "    df['Sample'] = sample\n",
    "    \n",
    "    # check if SNV and not other variant\n",
    "    df = df[df['Type'] == 'SNV']\n",
    "    \n",
    "    # normal QC\n",
    "    df = df[(df['Frequency'] >= FREQ) & (df['Coverage'] >= COV) & (df['Average quality'] >= QUAL)]\n",
    "    \n",
    "    # forward reverse more than cut-off (keep to 0 unless for specific reasons)\n",
    "    df = df[(df['Forward/reverse balance'] > FRB)]\n",
    "    \n",
    "    # exclude resistant genes\n",
    "    df = df[~df['Reference Position'].isin(DB_DR)]\n",
    "    \n",
    "    # exclude repeats\n",
    "    df = df[~df['Reference Position'].isin(DB_REP_NE)]\n",
    "    \n",
    "    # exclude exons\n",
    "    df = df[df['Overlapping annotations'].notna()]\n",
    "    df = df[~df['Amino acid change'].notna()]\n",
    "    \n",
    "    # exlude proximity snps\n",
    "    arr_sorted = np.sort(df['Reference Position'])\n",
    "    diff_next = np.diff(arr_sorted)\n",
    "    too_close = diff_next <= 50\n",
    "    mask = np.ones_like(arr_sorted, dtype=bool)\n",
    "    mask[:-1] &= ~too_close\n",
    "    mask[1:]  &= ~too_close\n",
    "    df = df[df['Reference Position'].isin(arr_sorted[mask])]\n",
    "    \n",
    "    # add all reference alleles to ref_dict\n",
    "    for index,row in df.iterrows():\n",
    "        ref_dict[row['Reference Position']] = row['Reference']\n",
    "    snp_dfs_filtered[sample] = df\n",
    "\n",
    "# order dictionary\n",
    "ref_dict = dict(sorted(ref_dict.items()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd69f93e",
   "metadata": {},
   "source": [
    "### Create fasta file used for alignment and phylogenetic analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "10d523de",
   "metadata": {},
   "outputs": [],
   "source": [
    "ffasta = open(OUTPUT+'snps.fasta','w')\n",
    "\n",
    "# create reference for tree root\n",
    "ffasta.write('>MTB\\n'+''.join(ref_dict.values())+'\\n')\n",
    "\n",
    "# read in from filtered dataframes\n",
    "for sample in snp_dfs_filtered:\n",
    "    df = snp_dfs_filtered[sample]\n",
    "    sample_dict = dict(zip(df['Reference Position'],df['Allele']))\n",
    "    missing_keys = set(ref_dict.keys()) - set(sample_dict.keys())\n",
    "    for key in missing_keys:\n",
    "        sample_dict[key] = ref_dict[key]\n",
    "    sample_dict = dict(sorted(sample_dict.items()))\n",
    "    seq = ''.join(sample_dict.values())\n",
    "    ffasta.write('>'+sample+'\\n'+str(seq)+'\\n')\n",
    "ffasta.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f29863f3",
   "metadata": {},
   "source": [
    "### Assign lineages using unfiltered dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "37c5ed9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variables\n",
    "\n",
    "LIN_FREQ = 20\n",
    "d_lin_name_number = dict(zip(DB_LIN['Lineage_number'],DB_LIN['Lineage_name']))\n",
    "\n",
    "# create a lineage matrix that can later be used for viz if needed\n",
    "lin_mat = pd.DataFrame()\n",
    "for sample in snp_dfs:\n",
    "    df = snp_dfs[sample]\n",
    "    df = df.set_index('Reference Position')\n",
    "    df = df.join(df_lin,rsuffix='_lin')\n",
    "    df = df[df['Allele'] == df['Allele_lin']]\n",
    "    df = df[df['Frequency'] >= LIN_FREQ]\n",
    "    df = df.groupby('Lineage_number').mean('Fequency')\n",
    "    for i in df.index:\n",
    "        lin_mat.at[i,sample] = df.at[i,'Frequency']\n",
    "\n",
    "# add lineage4.9 for samples that did not call any lineage\n",
    "for sample in snp_dfs:\n",
    "    if sample not in lin_mat.columns:\n",
    "        lin_mat.at['lineage4.9',sample] = 100\n",
    "        \n",
    "df = lin_mat.reset_index()\n",
    "df = df.rename(columns={'index':'Lineage Number'})\n",
    "df['Lineage Name'] = df['Lineage Number'].map(d_lin_name_number)\n",
    "df = pd.melt(df,id_vars=['Lineage Name','Lineage Number'])\n",
    "df = df[df['value'].notna()]\n",
    "df = df.rename(columns={'variable':'Sample','value':'Frequency'})\n",
    "df = df[['Sample','Lineage Name','Lineage Number','Frequency']]\n",
    "df.to_excel(OUTPUT+'/lineages.xlsx')\n",
    "        \n",
    "### BUILD FILE FOR LINEAGE ANNOTATION ON PHYLOGENETIC TREE ##\n",
    "\n",
    "# create lineage dictionary from lineage matrix for iTOL tree       \n",
    "lineage_tree = dict()\n",
    "for col in lin_mat.columns:\n",
    "    df = lin_mat[[col]].dropna()\n",
    "    deep_lin = max(df.index.to_list(), key=len)\n",
    "    lineage_tree[col] = deep_lin\n",
    "\n",
    "# create dictionary for colors\n",
    "colors = dict(zip(DB_LIN['Lineage_number'],DB_LIN['Lineage_color']))\n",
    "\n",
    "# build iTOL file for lineage\n",
    "flineage = open(OUTPUT+'lineage.txt','w')\n",
    "flineage.write('DATASET_COLORSTRIP\\nSEPARATOR TAB\\nDATASET_LABEL\\tLineage\\nCOLOR\\t#ff0000\\n')\n",
    "flineage.write('LEGEND_COLORS\\t'+'\\t'.join(colors.values())+'\\n')\n",
    "flineage.write('LEGEND_LABELS\\t'+'\\t'.join(colors.keys())+'\\n')\n",
    "\n",
    "flineage.write('\\nDATA\\n')\n",
    "for sample in lineage_tree:\n",
    "    flineage.write(sample+'\\t'+colors[lineage_tree[sample]]+'\\t'+lineage_tree[sample])\n",
    "    flineage.write('\\n')\n",
    "flineage.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d95b2a74",
   "metadata": {},
   "source": [
    "# RESISTANCE MODULE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "380d6b62",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: output/Resistance: File exists\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "256"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system('mkdir '+OUTPUT+'Resistance')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "014ab9cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "WHORules = {'katG':   ['Isoniazid'],\n",
    "            'pncA':   ['Pyrazinamide'],\n",
    "            'ddn':    ['Pyrazinamide','Delamanid'],\n",
    "            'fbiA':   ['Pyrazinamide','Delamanid'],\n",
    "            'fbiB':   ['Pyrazinamide','Delamanid'],\n",
    "            'fbiC':   ['Pyrazinamide','Delamanid'],\n",
    "            'fgd1':   ['Pyrazinamide','Delamanid'],\n",
    "            'Rv2983': ['Pyrazinamide','Delamanid'],\n",
    "            'gid':    ['Streptomycin'],\n",
    "            'ethA':   ['Prothionamide','Ethambutol'],\n",
    "            'tlyA':   ['Capreomycin'],\n",
    "            'pepQ' :  ['Bedaquiline','Clofazimine']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ef7c51de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# columns to select from files\n",
    "SNP_COLUMNS = ['Type','Reference Position','Reference','Allele','Count','Coverage','Frequency',\n",
    "               'Forward/reverse balance','Average quality','Coding region change','Amino acid change',\n",
    "               'Overlapping annotations']\n",
    "COV_COLUMNS = ['Name','Target region length','Mean coverage','Zero coverage bases']\n",
    "STAT_COLUMNS = ['Mapped reads','Average coverage']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "460ac259",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/k2259885/anaconda3/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Conditional Formatting extension is not supported and will be removed\n",
      "  warn(msg)\n",
      "/Users/k2259885/anaconda3/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Conditional Formatting extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "# parse databases\n",
    "DR_WHO = pd.read_excel('../Resistance/WHO-UCN-TB-2023.5-eng.xlsx',skiprows=2)\n",
    "DR_WHO = DR_WHO[['drug','gene','mutation','variant','tier',\n",
    "                'Present_SOLO_SR','Present_SOLO_R','Present_SOLO_S','FINAL CONFIDENCE GRADING']]\n",
    "DR_CTB = pd.read_excel('../Resistance/CTB_Mutation_Catalogue.xlsx')\n",
    "DR_CTB.columns = ['CTB_variant','CTB_drug','CTB_FINAL_CONFIDENCE_GRADING']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fab8f587",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# create dictionaries to use for resistotyping\n",
    "dCTB = defaultdict(dict)\n",
    "for index,row in DR_CTB.iterrows():\n",
    "    dCTB[row['CTB_variant']][row['CTB_drug']] = row['CTB_FINAL_CONFIDENCE_GRADING']\n",
    "dWHO = defaultdict(dict)\n",
    "for index,row in DR_WHO.iterrows():\n",
    "    dWHO[row['variant']][row['drug']] = row['FINAL CONFIDENCE GRADING']\n",
    "\n",
    "dGene2Drug = defaultdict(list)\n",
    "dDrug2Gene = defaultdict(list)\n",
    "\n",
    "for index,row in DR_WHO.iterrows():\n",
    "    dGene2Drug[row['gene']].append(row['drug'])\n",
    "    dDrug2Gene[row['drug']].append(row['gene'])\n",
    "for index,row in DR_CTB.iterrows():\n",
    "    dGene2Drug[row['CTB_variant'].split('_')[0]].append(row['CTB_drug'])\n",
    "    dDrug2Gene[row['CTB_drug']].append(row['CTB_variant'].split('_')[0])\n",
    "\n",
    "# set lists\n",
    "for gene in dGene2Drug:\n",
    "    dGene2Drug[gene] = list(set(dGene2Drug[gene]))\n",
    "for drug in dDrug2Gene:\n",
    "    dDrug2Gene[drug] = list(set(dDrug2Gene[drug]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b465e4aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all drug resistant genes\n",
    "DR_GENES = list(DR_WHO['gene'].unique())\n",
    "DR_GENES = list(set(DR_GENES + list(DR_CTB['CTB_variant'].str.split('_').str[0].unique())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d1289f79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'Amikacin': ['whiB7',\n",
       "              'rrs',\n",
       "              'ccsA',\n",
       "              'bacA',\n",
       "              'Rv2477c',\n",
       "              'whiB6',\n",
       "              'eis'],\n",
       "             'Bedaquiline': ['pepQ',\n",
       "              'mtrA',\n",
       "              'Rv0678',\n",
       "              'lpqB',\n",
       "              'Rv1979c',\n",
       "              'mmpL5',\n",
       "              'mmpS5',\n",
       "              'mtrB',\n",
       "              'atpE'],\n",
       "             'Capreomycin': ['rrs',\n",
       "              'ccsA',\n",
       "              'bacA',\n",
       "              'Rv2680',\n",
       "              'whiB6',\n",
       "              'rrl',\n",
       "              'Rv2681',\n",
       "              'tlyA'],\n",
       "             'Clofazimine': ['fbiA',\n",
       "              'pepQ',\n",
       "              'fbiC',\n",
       "              'Rv0678',\n",
       "              'mmpS5',\n",
       "              'Rv1979c',\n",
       "              'mmpL5',\n",
       "              'Rv2983',\n",
       "              'fgd1',\n",
       "              'fbiB'],\n",
       "             'Delamanid': ['fbiA',\n",
       "              'fbiC',\n",
       "              'ddn',\n",
       "              'Rv2983',\n",
       "              'fgd1',\n",
       "              'ndh',\n",
       "              'fbiB'],\n",
       "             'Ethambutol': ['iniB',\n",
       "              'embR',\n",
       "              'aftB',\n",
       "              'iniC',\n",
       "              'Rv2477c',\n",
       "              'iniA',\n",
       "              'ubiA',\n",
       "              'Rv2752c',\n",
       "              'eispromoter',\n",
       "              'rmlD',\n",
       "              'embB',\n",
       "              'embA',\n",
       "              'embC',\n",
       "              'glpK',\n",
       "              'eis'],\n",
       "             'Ethionamide': ['ethA',\n",
       "              'Rv0565c',\n",
       "              'mshA',\n",
       "              'inhApromoter',\n",
       "              'inhA',\n",
       "              'Rv3083',\n",
       "              'ethR',\n",
       "              'ndh'],\n",
       "             'Isoniazid': ['fabG1',\n",
       "              'hadA',\n",
       "              'Rv1258c',\n",
       "              'katG',\n",
       "              'kasA',\n",
       "              'Rv1129c',\n",
       "              'Rv2752c',\n",
       "              'mshA',\n",
       "              'inhApromoter',\n",
       "              'inhA',\n",
       "              'ndh',\n",
       "              'dnaA',\n",
       "              'Rv0010c',\n",
       "              'ahpC',\n",
       "              'glpK',\n",
       "              'ahpCpromoter'],\n",
       "             'Kanamycin': ['whiB7',\n",
       "              'rrs',\n",
       "              'ccsA',\n",
       "              'bacA',\n",
       "              'Rv2477c',\n",
       "              'whiB6',\n",
       "              'eis'],\n",
       "             'Levofloxacin': ['Rv1129c',\n",
       "              'Rv2477c',\n",
       "              'gyrA',\n",
       "              'Rv2752c',\n",
       "              'gyrB',\n",
       "              'glpK'],\n",
       "             'Linezolid': ['rplC', 'rrl', 'tsnR'],\n",
       "             'Moxifloxacin': ['Rv1129c',\n",
       "              'Rv2477c',\n",
       "              'gyrA',\n",
       "              'Rv2752c',\n",
       "              'gyrB',\n",
       "              'glpK'],\n",
       "             'Pyrazinamide': ['Rv1258c',\n",
       "              'sigE',\n",
       "              'pncApromoter',\n",
       "              'pncA',\n",
       "              'PPE35',\n",
       "              'rpsA',\n",
       "              'panD',\n",
       "              'Rv3236c',\n",
       "              'clpC1'],\n",
       "             'Rifampicin': ['rpoA',\n",
       "              'mtrA',\n",
       "              'Rv1129c',\n",
       "              'Rv2477c',\n",
       "              'lpqB',\n",
       "              'Rv2752c',\n",
       "              'rpoC',\n",
       "              'rpoB',\n",
       "              'glpK',\n",
       "              'mtrB',\n",
       "              'nusG'],\n",
       "             'Streptomycin': ['whiB7',\n",
       "              'Rv1258c',\n",
       "              'rrs',\n",
       "              'gid',\n",
       "              'bacA',\n",
       "              'rpsL',\n",
       "              'Rv2477c',\n",
       "              'glpK',\n",
       "              'gidB'],\n",
       "             'Aminoglycoside': ['thyA', 'eispromoter', 'rrs', 'tlyA'],\n",
       "             'Fluoroquinolone': ['gyrB'],\n",
       "             'Ethamobutol': ['manB'],\n",
       "             'Rifamycin': ['rpoB']})"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dDrug2Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ea99f8a9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'bacA': ['Kanamycin', 'Streptomycin', 'Capreomycin', 'Amikacin'],\n",
       "             'ccsA': ['Kanamycin', 'Capreomycin', 'Amikacin'],\n",
       "             'eis': ['Kanamycin', 'Ethambutol', 'Amikacin'],\n",
       "             'rrs': ['Kanamycin',\n",
       "              'Streptomycin',\n",
       "              'Capreomycin',\n",
       "              'Amikacin',\n",
       "              'Aminoglycoside'],\n",
       "             'Rv2477c': ['Kanamycin',\n",
       "              'Rifampicin',\n",
       "              'Levofloxacin',\n",
       "              'Streptomycin',\n",
       "              'Moxifloxacin',\n",
       "              'Ethambutol',\n",
       "              'Amikacin'],\n",
       "             'whiB6': ['Kanamycin', 'Capreomycin', 'Amikacin'],\n",
       "             'whiB7': ['Kanamycin', 'Streptomycin', 'Amikacin'],\n",
       "             'atpE': ['Bedaquiline'],\n",
       "             'lpqB': ['Bedaquiline', 'Rifampicin'],\n",
       "             'mmpL5': ['Bedaquiline', 'Clofazimine'],\n",
       "             'mmpS5': ['Bedaquiline', 'Clofazimine'],\n",
       "             'mtrA': ['Bedaquiline', 'Rifampicin'],\n",
       "             'mtrB': ['Bedaquiline', 'Rifampicin'],\n",
       "             'pepQ': ['Bedaquiline', 'Clofazimine'],\n",
       "             'Rv0678': ['Bedaquiline', 'Clofazimine'],\n",
       "             'Rv1979c': ['Bedaquiline', 'Clofazimine'],\n",
       "             'rrl': ['Linezolid', 'Capreomycin'],\n",
       "             'Rv2680': ['Capreomycin'],\n",
       "             'Rv2681': ['Capreomycin'],\n",
       "             'tlyA': ['Aminoglycoside', 'Capreomycin'],\n",
       "             'fbiA': ['Delamanid', 'Clofazimine'],\n",
       "             'fbiB': ['Delamanid', 'Clofazimine'],\n",
       "             'fbiC': ['Delamanid', 'Clofazimine'],\n",
       "             'fgd1': ['Delamanid', 'Clofazimine'],\n",
       "             'Rv2983': ['Delamanid', 'Clofazimine'],\n",
       "             'ddn': ['Delamanid'],\n",
       "             'ndh': ['Delamanid', 'Isoniazid', 'Ethionamide'],\n",
       "             'aftB': ['Ethambutol'],\n",
       "             'embA': ['Ethambutol'],\n",
       "             'embB': ['Ethambutol'],\n",
       "             'embC': ['Ethambutol'],\n",
       "             'embR': ['Ethambutol'],\n",
       "             'glpK': ['Rifampicin',\n",
       "              'Levofloxacin',\n",
       "              'Streptomycin',\n",
       "              'Moxifloxacin',\n",
       "              'Ethambutol',\n",
       "              'Isoniazid'],\n",
       "             'Rv2752c': ['Rifampicin',\n",
       "              'Levofloxacin',\n",
       "              'Moxifloxacin',\n",
       "              'Ethambutol',\n",
       "              'Isoniazid'],\n",
       "             'ubiA': ['Ethambutol'],\n",
       "             'ethA': ['Ethionamide'],\n",
       "             'ethR': ['Ethionamide'],\n",
       "             'inhA': ['Isoniazid', 'Ethionamide'],\n",
       "             'mshA': ['Isoniazid', 'Ethionamide'],\n",
       "             'Rv0565c': ['Ethionamide'],\n",
       "             'Rv3083': ['Ethionamide'],\n",
       "             'ahpC': ['Isoniazid'],\n",
       "             'dnaA': ['Isoniazid'],\n",
       "             'hadA': ['Isoniazid'],\n",
       "             'katG': ['Isoniazid'],\n",
       "             'Rv0010c': ['Isoniazid'],\n",
       "             'Rv1129c': ['Rifampicin',\n",
       "              'Moxifloxacin',\n",
       "              'Isoniazid',\n",
       "              'Levofloxacin'],\n",
       "             'Rv1258c': ['Pyrazinamide', 'Isoniazid', 'Streptomycin'],\n",
       "             'gyrA': ['Moxifloxacin', 'Levofloxacin'],\n",
       "             'gyrB': ['Moxifloxacin', 'Fluoroquinolone', 'Levofloxacin'],\n",
       "             'rplC': ['Linezolid'],\n",
       "             'tsnR': ['Linezolid'],\n",
       "             'clpC1': ['Pyrazinamide'],\n",
       "             'panD': ['Pyrazinamide'],\n",
       "             'pncA': ['Pyrazinamide'],\n",
       "             'PPE35': ['Pyrazinamide'],\n",
       "             'rpsA': ['Pyrazinamide'],\n",
       "             'Rv3236c': ['Pyrazinamide'],\n",
       "             'sigE': ['Pyrazinamide'],\n",
       "             'nusG': ['Rifampicin'],\n",
       "             'rpoA': ['Rifampicin'],\n",
       "             'rpoB': ['Rifamycin', 'Rifampicin'],\n",
       "             'rpoC': ['Rifampicin'],\n",
       "             'gid': ['Streptomycin'],\n",
       "             'rpsL': ['Streptomycin'],\n",
       "             'ahpCpromoter': ['Isoniazid'],\n",
       "             'eispromoter': ['Aminoglycoside', 'Ethambutol'],\n",
       "             'fabG1': ['Isoniazid'],\n",
       "             'gidB': ['Streptomycin'],\n",
       "             'inhApromoter': ['Isoniazid', 'Ethionamide'],\n",
       "             'iniA': ['Ethambutol'],\n",
       "             'iniB': ['Ethambutol'],\n",
       "             'iniC': ['Ethambutol'],\n",
       "             'kasA': ['Isoniazid'],\n",
       "             'manB': ['Ethamobutol'],\n",
       "             'pncApromoter': ['Pyrazinamide'],\n",
       "             'rmlD': ['Ethambutol'],\n",
       "             'thyA': ['Aminoglycoside']})"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dGene2Drug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "16b30373",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "aa64f5ba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# merge into single dataframe\n",
    "\n",
    "# to do --> any quality control to be included in resistotyping\n",
    "# to do --> inhA promoter issue\n",
    "\n",
    "DR_QUAL = 20\n",
    "DR_FRB = 0\n",
    "DR_FREQ = 5\n",
    "DR_COV = 5\n",
    "\n",
    "dRes = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "columns = []\n",
    "for drug in dDrug2Gene:\n",
    "    columns.append( ('Resistotype',drug,'') )\n",
    "    \n",
    "queries = ['Coding region change','Amino acid change','Resistance level']\n",
    "for drug in dDrug2Gene:\n",
    "    for gene in dDrug2Gene[drug]:\n",
    "        for query in queries:\n",
    "            columns.append( (drug,gene,query) )\n",
    "            \n",
    "columns = pd.MultiIndex.from_tuples(columns)\n",
    "mainOut = pd.DataFrame(columns=columns)\n",
    "\n",
    "for file in snp_files:\n",
    "    indexes = []\n",
    "    sample = file.replace('.snp','')\n",
    "    dfResSample = pd.DataFrame(columns=['Drug','Gene','Coding region change','Amino acid change','Level'])\n",
    "    counter = 0\n",
    "    \n",
    "    # read in files and do QC \n",
    "    snp = pd.read_csv('input/snp/'+sample+'.snp')[SNP_COLUMNS]\n",
    "    snp = snp[snp['Average quality'] >= DR_QUAL]\n",
    "    snp = snp[snp['Forward/reverse balance'] > DR_FRB]\n",
    "    snp = snp[snp['Frequency'] >= DR_FREQ]\n",
    "    snp = snp[snp['Coverage'] >= DR_COV]\n",
    "    \n",
    "    cov = pd.read_csv('input/cov/'+sample+'.cov')[COV_COLUMNS]\n",
    "    stat = pd.read_csv('input/stat/'+sample+'.stat')[STAT_COLUMNS]\n",
    "    \n",
    "    # split column to get gene\n",
    "    df = snp.copy()\n",
    "    df = df[df['Overlapping annotations'].notna()]\n",
    "    df = df[df['Overlapping annotations'].str.contains('Gene')]\n",
    "    df['Gene'] = df['Overlapping annotations'].str.split('Gene: ').str[1].str.split().str[0]\n",
    "    \n",
    "    # get only DR genes\n",
    "    df = df[df['Gene'].isin(DR_GENES)]\n",
    "    # merge snp and cov\n",
    "    df = pd.merge(df, cov, left_on='Gene', right_on='Name')\n",
    "    df = df.drop(columns=['Gene'])\n",
    "    # merge with stat\n",
    "    for col in stat.columns:\n",
    "        df[col] = stat.at[0,col]\n",
    "    \n",
    "    # add sample name as a column and move it to first column\n",
    "    df['Sample'] = sample\n",
    "    df = df[[df.columns[-1]] + df.columns[:-1].tolist()]\n",
    "    \n",
    "    # create resistance dictionary\n",
    "    \n",
    "    # check for ungraded mutations -  add ungraded for all genes associated with resistance\n",
    "    for index,row in df.iterrows():\n",
    "        if row['Name'] in dGene2Drug:\n",
    "            for drug in dGene2Drug[row['Name']]:\n",
    "                dfResSample.loc[len(dfResSample)] = [drug,str(row['Name']),str(row['Coding region change']),\n",
    "                                                     str(row['Amino acid change']),'Ungraded']\n",
    "                if 'Ungraded' not in dRes[sample][drug]:\n",
    "                    dRes[sample][drug].append('Ungraded')\n",
    "                    indexes.append(index)\n",
    "                 \n",
    "    # for ctb\n",
    "    df['tmp1'] = df['Name']+'_'+df['Coding region change'].str.split('c.').str[1] \n",
    "    df['tmp2'] = df['Name']+'_'+df['Amino acid change'].str.split('p.').str[1]\n",
    "    for index,row in df.iterrows():\n",
    "        if row['tmp1'] in dCTB:\n",
    "            for drug in dCTB[row['tmp1']]:\n",
    "                dRes[sample][drug].append(dCTB[row['tmp1']][drug])\n",
    "                indexes.append(index)\n",
    "                dfResSample.loc[len(dfResSample)] = [drug,row['Name'],str(row['tmp1']),\n",
    "                                                     str(row['tmp2']),dCTB[row['tmp1']][drug]]\n",
    "        if row['tmp2'] in dCTB:\n",
    "            for drug in dCTB[row['tmp2']]:\n",
    "                dRes[sample][drug].append(dCTB[row['tmp2']][drug])\n",
    "                indexes.append(index)\n",
    "                dfResSample.loc[len(dfResSample)] = [drug,row['Name'],str(row['tmp1']),\n",
    "                                                     str(row['tmp2']),dCTB[row['tmp2']][drug]]\n",
    "    df = df.drop(columns=['tmp1','tmp2'])\n",
    "    \n",
    "    # for WHO\n",
    "    df['tmp1'] = df['Name']+'_'+df['Coding region change'].str.split(':').str[1] \n",
    "    df['tmp2'] = df['Name']+'_'+df['Amino acid change'].str.split(':').str[1]\n",
    "    for index,row in df.iterrows():\n",
    "        if row['tmp1'] in dWHO:\n",
    "            for drug in dWHO[row['tmp1']]:\n",
    "                dRes[sample][drug].append(dWHO[row['tmp1']][drug])\n",
    "                indexes.append(index)\n",
    "                dfResSample.loc[len(dfResSample)] = [drug,row['Name'],str(row['tmp1']),\n",
    "                                                     str(row['tmp2']),dWHO[row['tmp1']][drug]]\n",
    "        if row['tmp2'] in dWHO:\n",
    "            for drug in dWHO[row['tmp2']]:\n",
    "                dRes[sample][drug].append(dWHO[row['tmp2']][drug])\n",
    "                indexes.append(index)\n",
    "                dfResSample.loc[len(dfResSample)] = [drug,row['Name'],str(row['tmp1']),\n",
    "                                                     str(row['tmp2']),dWHO[row['tmp2']][drug]]\n",
    "    df = df.drop(columns=['tmp1','tmp2'])\n",
    "    \n",
    "    # for WHO expert\n",
    "    df2 = df[(df['Type'].isin(['Insertion','Deletion'])) | \n",
    "             (df['Amino acid change'].str.contains('*',na=False,regex=False))]\n",
    "    for index,row in df2.iterrows():\n",
    "        if row['Name'] in WHORules:\n",
    "            for drug in WHORules[row['Name']]:\n",
    "                dRes[sample][drug].append('Expert_R')\n",
    "                dfResSample.loc[len(dfResSample)] = [drug,row['Name'],str(row['Coding region change']),\n",
    "                                                     str(row['Amino acid change']),'Expert_R']\n",
    "    if 'Rv0678' in list(df2['Name']):\n",
    "        if 'mmpL5' in list(df2['Name']):\n",
    "            pass\n",
    "        else:\n",
    "            dRes[sample]['Bedaquiline'].append('Expert_R')\n",
    "            dRes[sample]['Clofazimine'].append('Expert_R')\n",
    "            df3 = df2[df2['Name'] == 'Rv0678']\n",
    "            for index,row in df3.iterrows():\n",
    "                dfResSample.loc[len(dfResSample)] = ['Bedaquiline',row['Name'],\n",
    "                                                     str(row['Coding region change']),\n",
    "                                                     str(row['Amino acid change']),\n",
    "                                                     'Expert_R']\n",
    "                dfResSample.loc[len(dfResSample)] = ['Clofazimine',row['Name'],\n",
    "                                                     str(row['Coding region change']),\n",
    "                                                     str(row['Amino acid change']),\n",
    "                                                     'Expert_R']\n",
    "            \n",
    "    df = df.loc[list(set(indexes))]\n",
    "    mainOut = resistotyping(dfResSample,mainOut,sample)\n",
    "    \n",
    "dfRes = buildResistogram(dRes)\n",
    "for i in dfRes.index:\n",
    "    for c in dfRes.columns:\n",
    "        mainOut.at[i,('Resistotype',c,'')] = dfRes.at[i,c]\n",
    "mainOut = mainOut.dropna(axis=1, how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "8e477714",
   "metadata": {},
   "outputs": [],
   "source": [
    "mainOut.to_excel('test.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "c471ec43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is to have the resistogram also alone\n",
    "def buildResistogram(d):\n",
    "    dfRes = pd.DataFrame()\n",
    "    for sample in d:\n",
    "        for drug in d[sample]:\n",
    "            if '1) Assoc w R' in d[sample][drug]:\n",
    "                #dfRes.at[sample,drug] = '1'\n",
    "                dfRes.at[sample,drug] = 'AWR'\n",
    "            elif '2) Assoc w R - Interim' in d[sample][drug]:\n",
    "                #dfRes.at[sample,drug] = '2'\n",
    "                dfRes.at[sample,drug] = 'AWR-I'\n",
    "            elif 'Expert_R' in d[sample][drug]:\n",
    "                #dfRes.at[sample,drug] = '3'\n",
    "                dfRes.at[sample,drug] = 'WHO-E'\n",
    "            elif '3) Uncertain significance' in d[sample][drug]:\n",
    "                #dfRes.at[sample,drug] = '4'\n",
    "                dfRes.at[sample,drug] = 'US'\n",
    "            elif '5) Not assoc w R' in d[sample][drug]:\n",
    "                #dfRes.at[sample,drug] = '6'\n",
    "                dfRes.at[sample,drug] = 'NAWR'\n",
    "            elif '4) Not assoc w R - Interim' in d[sample][drug]:\n",
    "                #dfRes.at[sample,drug] = '7'\n",
    "                dfRes.at[sample,drug] = 'NAWR-I'\n",
    "            elif 'R' in d[sample][drug]:\n",
    "                #dfRes.at[sample,drug] = '5'\n",
    "                dfRes.at[sample,drug] = 'CTB-R'\n",
    "            elif 'Ungraded' in d[sample][drug]:\n",
    "                #dfRes.at[sample,drug] = 'Ungraded'\n",
    "                dfRes.at[sample,drug] = 'UNGR'\n",
    "    dfRes = dfRes.fillna('WT')\n",
    "    return dfRes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "94164db5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def resistotyping(df,outdf,sample):\n",
    "    for drug in df['Drug'].unique():\n",
    "        for gene in df['Gene'].unique():\n",
    "            df2 = df[(df['Drug'] == drug) & (df['Gene'] == gene)]\n",
    "            if '1) Assoc w R' in df2['Level']:\n",
    "                df2 = df2[df2['Level'] == '1) Assoc w R']\n",
    "            elif '2) Assoc w R - Interim' in df2['Level']:\n",
    "                df2 = df2[df2['Level'] == '2) Assoc w R - Interim']\n",
    "            elif 'Expert_R' in df2['Level']:\n",
    "                df2 = df2[df2['Level'] == 'Expert_R']\n",
    "            elif '3) Uncertain significance' in df2['Level']:\n",
    "                df2 = df2[df2['Level'] == '3) Uncertain significance']\n",
    "            elif '5) Not assoc w R' in df2['Level']:\n",
    "                df2 = df2[df2['Level'] == '5) Not assoc w R']\n",
    "            elif '4) Not assoc w R - Interim' in df2['Level']:\n",
    "                df2 = df2[df2['Level'] == '4) Not assoc w R - Interim']\n",
    "            elif 'R' in df2['Level']:\n",
    "                df2 = df2[df2['Level'] == 'R']\n",
    "            elif 'Ungraded' in df2['Level']:\n",
    "                df2 = df2[df2['Level'] == 'Ungraded']\n",
    "            \n",
    "            nucl_change = str(' / '.join(df2['Coding region change']))\n",
    "            aa_change = str(' / '.join(df2['Amino acid change']))\n",
    "    \n",
    "            if len(df2.index) > 0:\n",
    "                level = list(set(df2['Level']))[0]\n",
    "                # Ungraded will have several mutations (can remove if don't want all the mutations in there)\n",
    "                if level == 'Ungraded': \n",
    "                    outdf.at[sample,(drug,gene,'Coding region change')] = nucl_change\n",
    "                    outdf.at[sample,(drug,gene,'Amino acid change')] = aa_change\n",
    "                    outdf.at[sample,(drug,gene,'Resistance level')] = level\n",
    "                else:\n",
    "                    outdf.at[sample,(drug,gene,'Coding region change')] = nucl_change\n",
    "                    outdf.at[sample,(drug,gene,'Amino acid change')] = aa_change\n",
    "                    outdf.at[sample,(drug,gene,'Resistance level')] = level\n",
    "            \n",
    "    return outdf       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "626531fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"10\" halign=\"left\">Resistotype</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"3\" halign=\"left\">Fluoroquinolone</th>\n",
       "      <th colspan=\"3\" halign=\"left\">Ethamobutol</th>\n",
       "      <th colspan=\"3\" halign=\"left\">Rifamycin</th>\n",
       "      <th>Moxifloxacin</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Amikacin</th>\n",
       "      <th>Bedaquiline</th>\n",
       "      <th>Capreomycin</th>\n",
       "      <th>Clofazimine</th>\n",
       "      <th>Delamanid</th>\n",
       "      <th>Ethambutol</th>\n",
       "      <th>Ethionamide</th>\n",
       "      <th>Isoniazid</th>\n",
       "      <th>Kanamycin</th>\n",
       "      <th>Levofloxacin</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"3\" halign=\"left\">gyrB</th>\n",
       "      <th colspan=\"3\" halign=\"left\">manB</th>\n",
       "      <th colspan=\"3\" halign=\"left\">rpoB</th>\n",
       "      <th>iniA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>...</th>\n",
       "      <th>Coding region change</th>\n",
       "      <th>Amino acid change</th>\n",
       "      <th>Resistance level</th>\n",
       "      <th>Coding region change</th>\n",
       "      <th>Amino acid change</th>\n",
       "      <th>Resistance level</th>\n",
       "      <th>Coding region change</th>\n",
       "      <th>Amino acid change</th>\n",
       "      <th>Resistance level</th>\n",
       "      <th>Coding region change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TC00760619_S1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00856554_S51</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00764189</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00831960_S32</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00855201_S148</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Asp435Val / nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00862935_S189</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Asp435Val / nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00836558_1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Ser450Tyr / nan / rpoB_Ser450Tyr</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00777418_S47</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00832902_S78</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Asp435Val / nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00780707_S93</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00838386_S26</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 434 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Resistotype                                                 \n",
       "                   Amikacin Bedaquiline Capreomycin Clofazimine Delamanid   \n",
       "                                                                            \n",
       "TC00760619_S1           NaN         NaN         NaN         NaN       NaN  \\\n",
       "TC00856554_S51          NaN         NaN         NaN         NaN       NaN   \n",
       "TC00764189              NaN         NaN         NaN         NaN       NaN   \n",
       "TC00831960_S32          NaN         NaN         NaN         NaN       NaN   \n",
       "TC00855201_S148         NaN         NaN         NaN         NaN       NaN   \n",
       "TC00862935_S189         NaN         NaN         NaN         NaN       NaN   \n",
       "TC00836558_1            NaN         NaN         NaN         NaN       NaN   \n",
       "TC00777418_S47          NaN         NaN         NaN         NaN       NaN   \n",
       "TC00832902_S78          NaN         NaN         NaN         NaN       NaN   \n",
       "TC00780707_S93          NaN         NaN         NaN         NaN       NaN   \n",
       "TC00838386_S26          NaN         NaN         NaN         NaN       NaN   \n",
       "\n",
       "                                                                         ...   \n",
       "                Ethambutol Ethionamide Isoniazid Kanamycin Levofloxacin  ...   \n",
       "                                                                         ...   \n",
       "TC00760619_S1          NaN         NaN       NaN       NaN          NaN  ...  \\\n",
       "TC00856554_S51         NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00764189             NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00831960_S32         NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00855201_S148        NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00862935_S189        NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00836558_1           NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00777418_S47         NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00832902_S78         NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00780707_S93         NaN         NaN       NaN       NaN          NaN  ...   \n",
       "TC00838386_S26         NaN         NaN       NaN       NaN          NaN  ...   \n",
       "\n",
       "                     Fluoroquinolone                                      \n",
       "                                gyrB                                      \n",
       "                Coding region change Amino acid change Resistance level   \n",
       "TC00760619_S1                    NaN               NaN              NaN  \\\n",
       "TC00856554_S51                   NaN               NaN              NaN   \n",
       "TC00764189                       NaN               NaN              NaN   \n",
       "TC00831960_S32                   NaN               NaN              NaN   \n",
       "TC00855201_S148                  NaN               NaN              NaN   \n",
       "TC00862935_S189                  NaN               NaN              NaN   \n",
       "TC00836558_1                     NaN               NaN              NaN   \n",
       "TC00777418_S47                   NaN               NaN              NaN   \n",
       "TC00832902_S78                   NaN               NaN              NaN   \n",
       "TC00780707_S93                   NaN               NaN              NaN   \n",
       "TC00838386_S26                   NaN               NaN              NaN   \n",
       "\n",
       "                         Ethamobutol                                      \n",
       "                                manB                                      \n",
       "                Coding region change Amino acid change Resistance level   \n",
       "TC00760619_S1                    NaN               NaN              NaN  \\\n",
       "TC00856554_S51                   nan               NaN         Ungraded   \n",
       "TC00764189                       NaN               NaN              NaN   \n",
       "TC00831960_S32                   nan               NaN         Ungraded   \n",
       "TC00855201_S148                  NaN               NaN              NaN   \n",
       "TC00862935_S189                  NaN               NaN              NaN   \n",
       "TC00836558_1                     NaN               NaN              NaN   \n",
       "TC00777418_S47                   NaN               NaN              NaN   \n",
       "TC00832902_S78                   NaN               NaN              NaN   \n",
       "TC00780707_S93                   NaN               NaN              NaN   \n",
       "TC00838386_S26                   NaN               NaN              NaN   \n",
       "\n",
       "                                                      Rifamycin   \n",
       "                                                           rpoB   \n",
       "                                           Coding region change   \n",
       "TC00760619_S1    NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu  \\\n",
       "TC00856554_S51   NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu   \n",
       "TC00764189       NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu   \n",
       "TC00831960_S32   NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu   \n",
       "TC00855201_S148                   NP_215181.1:p.Asp435Val / nan   \n",
       "TC00862935_S189                   NP_215181.1:p.Asp435Val / nan   \n",
       "TC00836558_1     NP_215181.1:p.Ser450Tyr / nan / rpoB_Ser450Tyr   \n",
       "TC00777418_S47   NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu   \n",
       "TC00832902_S78                    NP_215181.1:p.Asp435Val / nan   \n",
       "TC00780707_S93   NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu   \n",
       "TC00838386_S26   NP_215181.1:p.Ser450Leu / nan / rpoB_Ser450Leu   \n",
       "\n",
       "                                                           Moxifloxacin  \n",
       "                                                                   iniA  \n",
       "                Amino acid change Resistance level Coding region change  \n",
       "TC00760619_S1                 NaN                R                       \n",
       "TC00856554_S51                NaN                R                  NaN  \n",
       "TC00764189                    NaN                R                  NaN  \n",
       "TC00831960_S32                NaN                R                  NaN  \n",
       "TC00855201_S148               NaN         Ungraded                  NaN  \n",
       "TC00862935_S189               NaN         Ungraded                  NaN  \n",
       "TC00836558_1                  NaN                R                  NaN  \n",
       "TC00777418_S47                NaN                R                  NaN  \n",
       "TC00832902_S78                NaN         Ungraded                  NaN  \n",
       "TC00780707_S93                NaN                R                  NaN  \n",
       "TC00838386_S26                NaN                R                  NaN  \n",
       "\n",
       "[11 rows x 434 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e42fded8",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6e3d183a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isoniazid</th>\n",
       "      <th>Levofloxacin</th>\n",
       "      <th>Moxifloxacin</th>\n",
       "      <th>Ethambutol</th>\n",
       "      <th>Clofazimine</th>\n",
       "      <th>Delamanid</th>\n",
       "      <th>Capreomycin</th>\n",
       "      <th>Amikacin</th>\n",
       "      <th>Kanamycin</th>\n",
       "      <th>Ethionamide</th>\n",
       "      <th>Rifamycin</th>\n",
       "      <th>Rifampicin</th>\n",
       "      <th>Bedaquiline</th>\n",
       "      <th>Streptomycin</th>\n",
       "      <th>Aminoglycoside</th>\n",
       "      <th>Pyrazinamide</th>\n",
       "      <th>Linezolid</th>\n",
       "      <th>Fluoroquinolone</th>\n",
       "      <th>Ethamobutol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TC00760619_S1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00856554_S51</th>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00764189</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00831960_S32</th>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00855201_S148</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00862935_S189</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>WT</td>\n",
       "      <td>Ungraded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00836558_1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00777418_S47</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00832902_S78</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00780707_S93</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TC00838386_S26</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Ungraded</td>\n",
       "      <td>Ungraded</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Isoniazid  Levofloxacin  Moxifloxacin  Ethambutol   \n",
       "TC00760619_S1          1.0           1.0           1.0         1.0  \\\n",
       "TC00856554_S51         4.0           6.0           6.0         1.0   \n",
       "TC00764189             1.0           1.0           1.0         1.0   \n",
       "TC00831960_S32         4.0           1.0           1.0         1.0   \n",
       "TC00855201_S148        1.0           1.0           1.0         1.0   \n",
       "TC00862935_S189        1.0           4.0           4.0         1.0   \n",
       "TC00836558_1           1.0           4.0           4.0         1.0   \n",
       "TC00777418_S47         1.0           1.0           1.0         1.0   \n",
       "TC00832902_S78         1.0           4.0           4.0         1.0   \n",
       "TC00780707_S93         1.0           1.0           1.0         1.0   \n",
       "TC00838386_S26         1.0           1.0           1.0         1.0   \n",
       "\n",
       "                 Clofazimine  Delamanid  Capreomycin  Amikacin Kanamycin   \n",
       "TC00760619_S1            4.0        7.0          6.0       4.0       4.0  \\\n",
       "TC00856554_S51           4.0        7.0          6.0       6.0       4.0   \n",
       "TC00764189               2.0        7.0          6.0  Ungraded  Ungraded   \n",
       "TC00831960_S32           6.0        7.0          6.0         6         4   \n",
       "TC00855201_S148          4.0        7.0          6.0  Ungraded  Ungraded   \n",
       "TC00862935_S189          4.0        7.0          6.0  Ungraded  Ungraded   \n",
       "TC00836558_1             2.0        7.0          6.0         4         4   \n",
       "TC00777418_S47           4.0        7.0          6.0  Ungraded  Ungraded   \n",
       "TC00832902_S78           4.0        7.0          6.0  Ungraded  Ungraded   \n",
       "TC00780707_S93           4.0        7.0          6.0  Ungraded  Ungraded   \n",
       "TC00838386_S26           4.0        7.0          6.0  Ungraded  Ungraded   \n",
       "\n",
       "                 Ethionamide Rifamycin  Rifampicin  Bedaquiline  Streptomycin   \n",
       "TC00760619_S1            4.0       5.0         1.0          4.0           4.0  \\\n",
       "TC00856554_S51           4.0       5.0         1.0          4.0           6.0   \n",
       "TC00764189               4.0       5.0         1.0          2.0           1.0   \n",
       "TC00831960_S32           4.0       5.0         1.0          4.0           1.0   \n",
       "TC00855201_S148          4.0  Ungraded         1.0          4.0           4.0   \n",
       "TC00862935_S189          4.0  Ungraded         1.0          4.0           4.0   \n",
       "TC00836558_1             4.0         5         1.0          1.0           6.0   \n",
       "TC00777418_S47           4.0         5         1.0          4.0           2.0   \n",
       "TC00832902_S78           4.0  Ungraded         1.0          4.0           4.0   \n",
       "TC00780707_S93           4.0         5         1.0          4.0           4.0   \n",
       "TC00838386_S26           4.0         5         1.0          4.0           4.0   \n",
       "\n",
       "                Aminoglycoside  Pyrazinamide  Linezolid Fluoroquinolone   \n",
       "TC00760619_S1         Ungraded           1.0        6.0              WT  \\\n",
       "TC00856554_S51        Ungraded           4.0        6.0        Ungraded   \n",
       "TC00764189            Ungraded           1.0        6.0              WT   \n",
       "TC00831960_S32        Ungraded           1.0        6.0        Ungraded   \n",
       "TC00855201_S148       Ungraded           1.0        6.0              WT   \n",
       "TC00862935_S189       Ungraded           1.0        6.0              WT   \n",
       "TC00836558_1          Ungraded           4.0        6.0              WT   \n",
       "TC00777418_S47        Ungraded           1.0        6.0        Ungraded   \n",
       "TC00832902_S78        Ungraded           1.0        6.0        Ungraded   \n",
       "TC00780707_S93        Ungraded           1.0        6.0              WT   \n",
       "TC00838386_S26        Ungraded           1.0        6.0        Ungraded   \n",
       "\n",
       "                Ethamobutol  \n",
       "TC00760619_S1            WT  \n",
       "TC00856554_S51     Ungraded  \n",
       "TC00764189               WT  \n",
       "TC00831960_S32     Ungraded  \n",
       "TC00855201_S148          WT  \n",
       "TC00862935_S189    Ungraded  \n",
       "TC00836558_1             WT  \n",
       "TC00777418_S47     Ungraded  \n",
       "TC00832902_S78     Ungraded  \n",
       "TC00780707_S93           WT  \n",
       "TC00838386_S26     Ungraded  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfRes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "cda05af5",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfRes.to_excel('output/Resistance/resistogram.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a531f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "    # perform WHO Expert ruling\n",
    "#    df = WHOExpert(df)\n",
    "    \n",
    "#     # merge with CTB mutations\n",
    "#     df_ctb = df.copy()\n",
    "#     df_ctb['tmp1'] = df_ctb['Name']+'_'+df_ctb['Coding region change'].str.split('c.').str[1] \n",
    "#     df_ctb['tmp2'] = df_ctb['Name']+'_'+df_ctb['Amino acid change'].str.split('p.').str[1]\n",
    "#     merged1 = pd.merge(df_ctb, DR_CTB, left_on='tmp1',right_on='CTB_variant',how='left',suffixes=('','_from_tmp1'))\n",
    "#     tmp1_na_mask = merged1['CTB_variant'].isna()\n",
    "#     merged2 = pd.merge(df_ctb[tmp1_na_mask],DR_CTB,left_on='tmp2',right_on='CTB_variant',how='left',suffixes=('', '_from_tmp2'))\n",
    "#     df_ctb = pd.concat([merged1[~tmp1_na_mask], merged2], ignore_index=True)\n",
    "#     df_ctb = df_ctb.drop(columns=['tmp1','tmp2'])\n",
    "    \n",
    "#     # merge with WHO mutations\n",
    "#     df_who = df.copy()\n",
    "#     df_who['tmp1'] = df_who['Name']+'_'+df_who['Coding region change'].str.split(':').str[1] \n",
    "#     df_who['tmp2'] = df_who['Name']+'_'+df_who['Amino acid change'].str.split(':').str[1]\n",
    "#     merged1 = pd.merge(df_who, DR_WHO, left_on='tmp1',right_on='variant',how='left',suffixes=('','_from_tmp1'))\n",
    "#     tmp1_na_mask = merged1['variant'].isna()\n",
    "#     merged2 = pd.merge(df_who[tmp1_na_mask],DR_WHO,left_on='tmp2',right_on='variant',how='left',suffixes=('', '_from_tmp2'))\n",
    "#     df_who = pd.concat([merged1[~tmp1_na_mask], merged2], ignore_index=True)\n",
    "#     df_who = df_who.drop(columns=['tmp1','tmp2'])\n",
    "    \n",
    "#     # create single file for sample resistance and write to excel file\n",
    "#     df = pd.concat([df_ctb,df_who])\n",
    "#     # select variants with DR annotation in CTB or WHO\n",
    "#     df = df[(df['CTB_FINAL_CONFIDENCE_GRADING'].notna()) | (df['FINAL CONFIDENCE GRADING'].notna())]\n",
    "    \n",
    "    #df.to_excel(OUTPUT+'Resistance/'+sample+'_RES.xlsx', index=False)\n",
    "    \n",
    "#     #resistotype = resistotyping(df)\n",
    "#     #for drug in resistotype:\n",
    "#     #    print(str(drug) + '\\t' + str(resistotype[drug]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "b346d59e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Reference Position</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Count</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Frequency</th>\n",
       "      <th>Forward/reverse balance</th>\n",
       "      <th>Average quality</th>\n",
       "      <th>Coding region change</th>\n",
       "      <th>Amino acid change</th>\n",
       "      <th>Overlapping annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SNV</td>\n",
       "      <td>1849</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>182</td>\n",
       "      <td>182</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.492891</td>\n",
       "      <td>33.670330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SNV</td>\n",
       "      <td>1977</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>157</td>\n",
       "      <td>158</td>\n",
       "      <td>99.367089</td>\n",
       "      <td>0.456522</td>\n",
       "      <td>33.707006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>SNV</td>\n",
       "      <td>4013</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>137</td>\n",
       "      <td>137</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.469512</td>\n",
       "      <td>33.182482</td>\n",
       "      <td>NP_214517.1:c.734T&gt;C</td>\n",
       "      <td>NP_214517.1:p.Ile245Thr</td>\n",
       "      <td>CDS: recF, Gene: recF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SNV</td>\n",
       "      <td>7362</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>110</td>\n",
       "      <td>112</td>\n",
       "      <td>98.214286</td>\n",
       "      <td>0.393443</td>\n",
       "      <td>33.054545</td>\n",
       "      <td>NP_214520.1:c.61G&gt;C</td>\n",
       "      <td>NP_214520.1:p.Glu21Gln</td>\n",
       "      <td>CDS: gyrA, Gene: gyrA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>SNV</td>\n",
       "      <td>7570</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>139</td>\n",
       "      <td>139</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.406250</td>\n",
       "      <td>32.604317</td>\n",
       "      <td>NP_214520.1:c.269C&gt;T</td>\n",
       "      <td>NP_214520.1:p.Ala90Val</td>\n",
       "      <td>CDS: gyrA, Gene: gyrA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55658</th>\n",
       "      <td>SNV</td>\n",
       "      <td>4407588</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>179</td>\n",
       "      <td>179</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>33.039106</td>\n",
       "      <td>NP_218436.2:c.615A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDS: gid, Gene: gid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55662</th>\n",
       "      <td>SNV</td>\n",
       "      <td>4407927</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>139</td>\n",
       "      <td>141</td>\n",
       "      <td>98.581560</td>\n",
       "      <td>0.494118</td>\n",
       "      <td>33.410072</td>\n",
       "      <td>NP_218436.2:c.276A&gt;C</td>\n",
       "      <td>NP_218436.2:p.Glu92Asp</td>\n",
       "      <td>CDS: gid, Gene: gid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55663</th>\n",
       "      <td>SNV</td>\n",
       "      <td>4407967</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.463158</td>\n",
       "      <td>33.013889</td>\n",
       "      <td>NP_218436.2:c.236T&gt;C</td>\n",
       "      <td>NP_218436.2:p.Leu79Ser</td>\n",
       "      <td>CDS: gid, Gene: gid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55669</th>\n",
       "      <td>SNV</td>\n",
       "      <td>4408899</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>135</td>\n",
       "      <td>136</td>\n",
       "      <td>99.264706</td>\n",
       "      <td>0.496894</td>\n",
       "      <td>32.888889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55670</th>\n",
       "      <td>SNV</td>\n",
       "      <td>4408923</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>138</td>\n",
       "      <td>138</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.414013</td>\n",
       "      <td>32.753623</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3096 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Type  Reference Position Reference Allele  Count  Coverage   Frequency   \n",
       "19     SNV                1849         C      A    182       182  100.000000  \\\n",
       "21     SNV                1977         A      G    157       158   99.367089   \n",
       "80     SNV                4013         T      C    137       137  100.000000   \n",
       "99     SNV                7362         G      C    110       112   98.214286   \n",
       "100    SNV                7570         C      T    139       139  100.000000   \n",
       "...    ...                 ...       ...    ...    ...       ...         ...   \n",
       "55658  SNV             4407588         T      C    179       179  100.000000   \n",
       "55662  SNV             4407927         T      G    139       141   98.581560   \n",
       "55663  SNV             4407967         A      G    144       144  100.000000   \n",
       "55669  SNV             4408899         G      A    135       136   99.264706   \n",
       "55670  SNV             4408923         C      T    138       138  100.000000   \n",
       "\n",
       "       Forward/reverse balance  Average quality  Coding region change   \n",
       "19                    0.492891        33.670330                   NaN  \\\n",
       "21                    0.456522        33.707006                   NaN   \n",
       "80                    0.469512        33.182482  NP_214517.1:c.734T>C   \n",
       "99                    0.393443        33.054545   NP_214520.1:c.61G>C   \n",
       "100                   0.406250        32.604317  NP_214520.1:c.269C>T   \n",
       "...                        ...              ...                   ...   \n",
       "55658                 0.500000        33.039106  NP_218436.2:c.615A>G   \n",
       "55662                 0.494118        33.410072  NP_218436.2:c.276A>C   \n",
       "55663                 0.463158        33.013889  NP_218436.2:c.236T>C   \n",
       "55669                 0.496894        32.888889                   NaN   \n",
       "55670                 0.414013        32.753623                   NaN   \n",
       "\n",
       "             Amino acid change Overlapping annotations  \n",
       "19                         NaN                     NaN  \n",
       "21                         NaN                     NaN  \n",
       "80     NP_214517.1:p.Ile245Thr   CDS: recF, Gene: recF  \n",
       "99      NP_214520.1:p.Glu21Gln   CDS: gyrA, Gene: gyrA  \n",
       "100     NP_214520.1:p.Ala90Val   CDS: gyrA, Gene: gyrA  \n",
       "...                        ...                     ...  \n",
       "55658                      NaN     CDS: gid, Gene: gid  \n",
       "55662   NP_218436.2:p.Glu92Asp     CDS: gid, Gene: gid  \n",
       "55663   NP_218436.2:p.Leu79Ser     CDS: gid, Gene: gid  \n",
       "55669                      NaN                     NaN  \n",
       "55670                      NaN                     NaN  \n",
       "\n",
       "[3096 rows x 12 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cecf0025",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdcdb0d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ba6d569",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0773fef7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1df0e0f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e22f345",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9f17741",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5636f638",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
